### ISSUE TITLE
discrepancies in smrybio and totbio reporting

### ISSUE THREAD

**Issue by Rick-Methot-NOAA** [nmfs-stock-synthesis/stock-synthesis#154]
Overall the calculation and storage of totbio and smrybio is messier and more convoluted than it should be.  Some are calculated on the fly, and some not until the write stage.  
The Time_Series table, which r4ss relies on, calculates then using:
     temp=natage(t,p,g)(Smry_Age,nages)*Save_Wt_Age(bio_t,g)(Smry_Age,nages);
     Bio_Comp(GP(g))+=value(temp);   //sums to accumulate across platoons and settlements
     Num_Comp(GP(g))+=value(sum(natage(t,p,g)(Smry_Age,nages)));
     totbio+= natage(t,p,g)*Save_Wt_Age(bio_t,g);
     smrybio+= temp;
     smrynum+=sum(natage(t,p,g)(Smry_Age,nages));
     smryage+=natage(t,p,g)(Smry_Age,nages)*r_ages(Smry_Age,nages);

But the values output for ss_summary.sso are calculated within ss_popdyn because smrybio needs to be available as a basis for density-dependent parameters.

Sometime after 3.30.16, the calculation within ss_popdyn went awry for purpose of smrybio reporting, but not for any other noticeable dependent purpose.

Also note that ss_summary was reporting smrybio for both smrybio and totbio for time_series years only; forecast years remained correct.

With this issue, the whole situation will be investigated and a consistent, tested approach implemented.

And note that

**Comment by k-doering-NOAA** [comment]
Some observations using the Simple model, in case they are helpful. I ran simple using 3.30.16 and then using the [current main branch](https://github.com/nmfs-stock-synthesis/stock-synthesis/tree/a923c2546dd9bff6aafd43c63bf2528a05d2f32e).

In 3.30.16:
- Totbio values (labeled AllBio) and SmryBio for all years (model years and forecast years) are different in Report.sso, but are the same in ss_summary.sso. It seems like SmryBio values are being used for both quantities in ss_summary.sso. However, Totbio virgin and init in ss_summary.sso match Report.sso.
-  SmryBio_virgin is 0 in ss_summary.sso, but is 124144 in Report.sso.

In current main: (tried to italicize some observations to highlight the differences between .16 and main)
- Totbio values (labeled AllBio) and SmryBio for *model years only* are different in Report.sso, but are the same in ss_summary.sso. In this case, the values are *different than any of the output in the Report.sso file*.However, Totbio virgin and init in ss_summary.sso match Report.sso.  (note Forecast values for both Totbio and SmryBio in ss_summary.sso are both the same as the Bio_summary in Report.sso, as observed in .16)
- SmryBio_virgin is 0 in ss_summary.sso, but is 124144 in Report.sso (as in 3.30.16).

This means .16 ss_summary.sso values are not a good basis of comparison, but perhaps the Report.sso values are? Values in the TIME_SERIES table are identical between 3.30.16 and the current main branch.

**Comment by Rick-Methot-NOAA** [comment]
fix original problem after finding that ALK was not getting updated with correct subseas.

Continue working after creating an example with time-vary_growth.  Transition between years is not being handled correctly.

**Comment by k-doering-NOAA** [comment]
Note there is an [example with timevarying growth using blocks](https://github.com/nmfs-stock-synthesis/ss-test-models/tree/main/models/growth_timevary)  in case it is helpful. Approx. run time is ~ 1 min

**Comment by Rick-Methot-NOAA** [comment]
This and issue #155 are both resolved.
A fuller accounting can be found with issue #155

**Comment by k-doering-NOAA** [comment]
There are still some small details to take care of with this issue:
1. Tot bio during model and forecast years in summary.sso are still matching the Smrybio in the report file rather than totbio... I wonder if just changing the column this line references to 1 would solve it? https://github.com/nmfs-stock-synthesis/stock-synthesis/blob/main/SS_write.tpl#L521
SS_smry<<"TotBio_"<<y<<" "<<Smry_Table(y,2)<<" 0.0"<<endl;
2. SmryBio_virgin is 0 in ss_summary.sso, but is 124144 in Report.sso. I took a look briefly at the code where report file and ss_summary.sso but find where the issue is occuring.

**Comment by Rick-Methot-NOAA** [comment]
My comparison between 30.16 and this new code version shows a very small (<1%) difference in the variance of spawnbiomass and recruitment for all forecast years.  No differences for earlier years.

**Comment by Rick-Methot-NOAA** [comment]
I suspected that this code in forecast is responsible for this small change in variance:
      join1=1./(1.+mfexp(30.*(Fcast_Fmult-max_harvest_rate)));
      Fcast_Fmult=join1*Fcast_Fmult + (1.-join1)*max_harvest_rate; // new F value for this fleet, constrained by max_harvest_rate
      if(join1<0.999)
      {
      	report5<<"Forecast F capped by max possible F from control file"<<max_harvest_rate<<endl;
     		N_warn++;  warning<<N_warn<<" Forecast F capped by max possible F from control file: "<<max_harvest_rate<<endl;
      }

But removal of that code did not restore the previous variance output.
Other possibility is the change to recr_dist associated with changes to improve its time-varying capability  So a comparison to the commit associated with storing recr_dist() with a year index would be the next test to do, but a low priority.

**Comment by Rick-Methot-NOAA** [comment]
Also find that the output for the "initial" year as reported in ss_summary was not the initial equilibrium because it was taken after the early recdevs were applied.  Verify that this is also now corrected.
